Source/DafnyCore/Compilers/SinglePassCompiler.cs

```diff
@@ -2650,10 +2650,11 @@ protected class NullClassWriter : IClassWriter {
         // finally, the jump back to the head of the function
         EmitJumpToTailCallStart(wr);
 
-      } else if (expr is BinaryExpr bin && bin.AccumulatesForTailRecursion != BinaryExpr.AccumulationOperand.None) {
-        Contract.Assert(accumulatorVar != null);
-        Contract.Assert(enclosingFunction != null);
-        Contract.Assert(enclosingFunction.IsAccumulatorTailRecursive);
```

Comment:
My interpretation of the purpose of these calls to `Contract.Assert` is to check properties that are expected to always hold if the surrounding conditional expression is true. Although adding them to the conditional will prevent a crash, I'm wondering whether you've done any deeper investigation into why they're not always true in this case. I worry that there could be a deeper bug lurking that this just temporarily hides away.

Reply:
Good question.
So here is the story as I understand it. The following code is a heuristic to check if a function is tail-recursive or not. It goes from bottom-up.
Look at `u.size() + 1` in the code.
For binary operators, if the left operation is tail-recursive (which it is `u.size()`), and if the right operand is `Triviallytailrecursive` (`1` is trivially tail recursive), then the binaryExpr's accumulator `AccumulatesForTailRecursion` is set to `BinaryExpr.AccumulationOperand.Right`, which makes sense. If the function is tail-recursive, the accumulation operand is the right one. But at this point, we haven't determined if the function is really tail-recursive or not.
https://github.com/dafny-lang/dafny/blob/8e2c073aac5f0cba4169324d0a02d99c1dd36d49/Source/DafnyCore/Resolver.cs#L7617-L7646

What makes no sense is, in the function `SinglePassCompiler.TrExprOpt`, to convert a binary expression to a tail-recursive equivalent only on the basis that `BinaryExpr.AccumulationOperand` is not `None`, because, as we've seen earlier, this was a conditional accumulation. There was another branch `l.size() + r.size()` that would eventually make the function not tail-recursive.

Reply:
I think that makes sense. It perhaps would be a good idea at some point to change the way the accumulation operand is set (e.g., have it be `None` if the function isn't tail recursive), but I think it clarifies that the additional condition seems like the right approach.